kernel/mem_domain.c

```diff
@@ -164,13 +164,18 @@ void k_mem_domain_add_thread(struct k_mem_domain *domain, k_tid_t thread)
 void k_mem_domain_remove_thread(k_tid_t thread)
 {
 	unsigned int key;
+	struct k_mem_domain *domain;
 
 	__ASSERT(thread && thread->mem_domain_info.mem_domain, "");
 
 	key = irq_lock();
+	domain = thread->mem_domain_info.mem_domain;
+	if (_current == thread) {
+		_arch_mem_domain_destroy(domain);
+	}
 
 	sys_dlist_remove(&thread->mem_domain_info.mem_domain_q_node);
-	thread->mem_domain_info.mem_domain = NULL;
```

Comment:
thread->mem_domain_info.mem_domain needs to be set to NULL for arch arm. (so does x86, I think.)
It seems the local variable 'domain' is unnecessary.

Reply:
The local variable is manual optimization. I feel instead of dereferencing it multiple times a single dereference should be better.
And coming to the functional part I will still set the mem_domain structure to NULL.

Reply:
Even you set domain = NULL, the thread->mem_domain_info.mem_domain is still keep the same value.
Or anything I misunderstand?

Reply:
> The local variable is manual optimization. I feel instead of dereferencing it multiple times a single dereference should be better.

Are you trying to out-smart the compiler?
At any rate, this looks wrong, pointer to the mem domain in the thread struct is NOT being cleared. Did you think that "domain = NULL" would somehow do this?

Reply:
oh my mistake I thought I did a double pointer here. will revert this line back.